### PR TITLE
switch to using QTime

### PR DESCRIPTION
**Pull request by ebbit1q**
I really liked your pull request so I made a pull request for your pull request so you can pull request while you pull request.

### CODE REVIEW

File: cockatrice/src/dlg_filter_games.cpp [hunk 1]
```diff
@@ -16,7 +16,13 @@
 DlgFilterGames::DlgFilterGames(const QMap<int, QString> &_allGameTypes,
                                const GamesProxyModel *_gamesProxyModel,
                                QWidget *parent)
-    : QDialog(parent), allGameTypes(_allGameTypes), gamesProxyModel(_gamesProxyModel)
+    : QDialog(parent), allGameTypes(_allGameTypes), gamesProxyModel(_gamesProxyModel),
+      gameAgeMap({{QTime(), tr("no limit")},
```
Comment (ebbit1q) [inline cockatrice/src/dlg_filter_games.cpp:20]: these can't be static for translation reasons so they are initialized here

File: cockatrice/src/dlg_filter_games.cpp [hunk 2]
```diff
@@ -16,7 +16,13 @@
...
+                  {QTime(0, 5), tr("5 minutes")},
```
Comment (ebbit1q) [inline cockatrice/src/dlg_filter_games.cpp:21]: I switched to not using the 5+ min because it looks really weird

File: cockatrice/src/dlg_filter_games.cpp [hunk 3]
```diff
@@ -230,9 +237,13 @@ int DlgFilterGames::getMaxPlayersFilterMax() const
     return maxPlayersFilterMaxSpinBox->value();
 }
 
-int DlgFilterGames::getMaxGameAgeComboBoxIndex() const
+const QTime &DlgFilterGames::getMaxGameAge() const
 {
-    return maxGameAgeComboBox->currentIndex();
+    int index = maxGameAgeComboBox->currentIndex();
+    if (index < 0 || index >= gameAgeMap.size()) { // index is out of bounds
+        return gamesProxyModel->getMaxGameAge();   // leave the setting unchanged
```
Comment (ebbit1q) [inline cockatrice/src/dlg_filter_games.cpp:244]: if an alternative method for setting this value will arise (or the file is edited) this value won't be in the list so it's set to -1, this shows as an empty entry, while I don't think this will come up soon I think it's good to not have it crash then.

File: cockatrice/src/gameselector.cpp [hunk 4]
```diff
@@ -81,8 +81,9 @@ GameSelector::GameSelector(AbstractClient *_client,
     if (room) {
         createButton = new QPushButton;
         connect(createButton, SIGNAL(clicked()), this, SLOT(actCreate()));
-    } else
-        createButton = 0;
+    } else {
+        createButton = nullptr;
```
Comment (ebbit1q) [inline cockatrice/src/gameselector.cpp:85]: unnecessary cosmetic change :upside_down_face:

File: cockatrice/src/gamesmodel.cpp [hunk 5]
```diff
@@ -66,15 +65,18 @@ QVariant GamesModel::data(const QModelIndex &index, int role) const
...
             return rooms.value(g.room_id());
         case CREATED: {
-            QDateTime then;
-            then.setTime_t(g.start_time());
-            int secs = then.secsTo(QDateTime::currentDateTime());
-
             switch (role) {
-                case Qt::DisplayRole:
+                case Qt::DisplayRole: {
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 8, 0))
```
Comment (ebbit1q) [inline cockatrice/src/gamesmodel.cpp:70]: while this is unrelated to @knitknit 's pull request, I think it's important to keep a bit of consistency

File: cockatrice/src/gamesmodel.cpp [hunk 6]
```diff
@@ -66,15 +65,18 @@ QVariant GamesModel::data(const QModelIndex &index, int role) const
...
+                    QDateTime then = QDateTime::fromSecsSinceEpoch(g.start_time(), Qt::UTC);
+#else
+                    QDateTime then = QDateTime::fromTime_t(g.start_time(), Qt::UTC);
+#endif
+                    int secs = then.secsTo(QDateTime::currentDateTimeUtc());
                     return getGameCreatedString(secs);
+                }
                 case SORT_ROLE:
-                    return QVariant(secs);
+                    return QVariant(-g.start_time());
```
Comment (ebbit1q) [inline cockatrice/src/gamesmodel.cpp]: unnecessary performance change :upside_down_face:

File: cockatrice/src/gamesmodel.cpp [hunk 7]
```diff
@@ -430,6 +414,11 @@ bool GamesProxyModel::filterAcceptsRow(int sourceRow, const QModelIndex & /*sour
 
 bool GamesProxyModel::filterAcceptsRow(int sourceRow) const
 {
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 8, 0))
+    static const QDate epochDate = QDateTime::fromSecsSinceEpoch(0, Qt::UTC).date();
+#else
+    static const QDate epochDate = QDateTime::fromTime_t(0, Qt::UTC).date();
```
Comment (ebbit1q) [inline cockatrice/src/gamesmodel.cpp:420]: this epochdate is used to check if the QTime value didn't wrap to the next day

File: cockatrice/src/gamesmodel.cpp [hunk 8]
```diff
@@ -472,10 +461,10 @@ bool GamesProxyModel::filterAcceptsRow(int sourceRow) const
     if ((int)game.max_players() > maxPlayersFilterMax)
         return false;
 
-    if (maxGameAgeSeconds != DEFAULT_MAX_GAME_AGE_SECONDS) {
-        QDateTime then;
-        then.setTime_t(game.start_time());
-        if (then.secsTo(QDateTime::currentDateTime()) > maxGameAgeSeconds) {
+    if (maxGameAge.isValid()) {
```
Comment (ebbit1q) [inline cockatrice/src/gamesmodel.cpp:464]: instead of checking if it's the default value, which not necessarily means it is nothing, we check if it's correctly formatted, an unset QTime() is invalid

File: cockatrice/src/gamesmodel.cpp [hunk 9]
```diff
@@ -472,10 +461,10 @@ bool GamesProxyModel::filterAcceptsRow(int sourceRow) const
...
+        QDateTime now = QDateTime::currentDateTimeUtc();
+        QDateTime total = now.addSecs(-(qint64)game.start_time()); // cast to 64 bit value before 2038-1-19 comes around
```
Comment (ebbit1q) [inline cockatrice/src/gamesmodel.cpp]: in 2038 this value will become 1<<31 and wrap to negative if it were cast to a plain int

File: cockatrice/src/gamesmodel.h [hunk 10]
```diff
@@ -11,6 +10,15 @@
...
 #include <QStringList>
+#include <QTime>
+
+const bool DEFAULT_UNAVAILABLE_GAMES_VISIBLE = false;
+const bool DEFAULT_SHOW_PASSWORD_PROTECTED_GAMES = true;
+const bool DEFAULT_SHOW_BUDDIES_ONLY_GAMES = true;
+const bool DEFAULT_HIDE_IGNORED_USER_GAMES = false;
+const int DEFAULT_MAX_PLAYERS_MIN = 1;
+const int DEFAULT_MAX_PLAYERS_MAX = 99;
+constexpr QTime DEFAULT_MAX_GAME_AGE = QTime();
```
Comment (ebbit1q) [inline cockatrice/src/gamesmodel.h:21]: I had to move these here, because it didn't like the static constexpr, I think it's fine?

File: cockatrice/src/gametimefilter.h [hunk 11]
```diff
@@ -1,65 +0,0 @@
...
-#include <QDebug>
-#include <QVector>
-
-// If this enum is ever expanded, make sure to update:
-// - getMaxGameAgeEnumOptions (order shown is order presented to user)
-// - text option generation in getMaxGameAgeInSecondsFromEnum
-// - validation conditions in getMaxGameAgeInSecondsFromComboBoxIndex
-enum MaxGameAge
-{
-    NO_MAX_AGE = 0,
```
Comment (ebbit1q) [inline cockatrice/src/gametimefilter.h:13]: I don't like enums :upside_down_face:

File: cockatrice/src/settings/gamefilterssettings.cpp [hunk 12]
```diff
@@ -92,15 +93,15 @@ int GameFiltersSettings::getMaxPlayers()
...
+    setValue(maxGameAge, "max_game_age_time", "filter_games");
 }
 
-int GameFiltersSettings::getMaxGameAgeSeconds()
+QTime GameFiltersSettings::getMaxGameAge()
 {
-    QVariant previous = getValue("max_game_age_seconds", "filter_games");
-    return previous == QVariant() ? -1 : previous.toInt();
+    QVariant previous = getValue("max_game_age_time", "filter_games");
+    return previous.toTime();
```
Comment (ebbit1q) [inline cockatrice/src/settings/gamefilterssettings.cpp:104]: this defaults to an invalid QTime() if the value doesn't match, so no harm in always using toTime()

**Review by ebbit1q — COMMENTED** [review]
here's some notes

File: cockatrice/src/gamesmodel.cpp (hunk 9 above)
Comment (ZeldaZach) [reply cockatrice/src/gamesmodel.cpp]: Lets be more careful about wrapping then. I'd consider changing the name of the variable then to indicate casting is a bad idea

File: cockatrice/src/gametimefilter.h (hunk 11 above)
Comment (ZeldaZach) [reply cockatrice/src/gametimefilter.h:13]: Really? Enums are great imo

File: cockatrice/src/dlg_filter_games.cpp (hunk 1 above)
Comment (ZeldaZach) [reply cockatrice/src/dlg_filter_games.cpp:20]: Seems fine

File: cockatrice/src/dlg_filter_games.cpp (hunk 2 above)
Comment (ZeldaZach) [reply cockatrice/src/dlg_filter_games.cpp:21]: Agreed

File: cockatrice/src/gamesmodel.cpp (hunk 6 above)
Comment (ZeldaZach) [reply cockatrice/src/gamesmodel.cpp]: Not sure what the point of this one is

File: cockatrice/src/gamesmodel.cpp [hunk 13]
```diff
@@ -365,7 +349,7 @@ void GamesProxyModel::resetFilterParameters()
     gameTypeFilter.clear();
     maxPlayersFilterMin = DEFAULT_MAX_PLAYERS_MIN;
     maxPlayersFilterMax = DEFAULT_MAX_PLAYERS_MAX;
-    maxGameAgeSeconds = DEFAULT_MAX_GAME_AGE_SECONDS;
+    maxGameAge = DEFAULT_MAX_GAME_AGE;
```
Comment (ZeldaZach) [inline cockatrice/src/gamesmodel.cpp:352]: Having a unit attached, especially for time, is important

Comment (ebbit1q) [reply cockatrice/src/gamesmodel.cpp:352]: qtime comprises units, you can access them with .hour() and .minute() for example

Comment (ZeldaZach) [reply cockatrice/src/gamesmodel.cpp:352]: Oh, gotcha

Comment (ebbit1q) [reply cockatrice/src/gamesmodel.cpp:352]: I had _TIME appended to it but it's already a QTime object so it felt a bit double

File: cockatrice/src/gametimefilter.h (hunk 11 above)
Comment (ebbit1q) [reply cockatrice/src/gametimefilter.h:13]: not for making changes later, consider you want to add 20 minutes in between while not breaking everyone's current settings

File: cockatrice/src/gamesmodel.cpp (hunk 6 above)
Comment (ebbit1q) [reply cockatrice/src/gamesmodel.cpp]: functionally the same, it just avoids calculating the difference with now, consider that this value is only used for sorting

Comment (ebbit1q) [reply cockatrice/src/gamesmodel.cpp]: actually, I'm not 100% on if this needs an extra cast to not break during 2038 suddenly

**Comment by ebbit1q** [comment]
the settings entry will look like this btw ```max_game_age_time=@Variant(\0\0\0\xf\0\x4\x93\xe0)``` it just encodes the data, but editing the ini files isn't really intended anyway

File: cockatrice/src/gamesmodel.cpp (hunk 9 above)
Comment (ebbit1q) [reply cockatrice/src/gamesmodel.cpp]: if you look at the if statement above you'll note we should consider never allowing more than 2147483647 players

**Review by knitknit — APPROVED** [review]
Awesome, thanks so much!
